docs/ref/forms/api.txt

```diff
@@ -1003,29 +1007,42 @@ method you're using:
     ...     "sender": "invalid email address",
     ...     "cc_myself": True,
     ... }
-    >>> f = ContactForm(data, auto_id=False)
+    >>> f = ContactForm(data)
     >>> print(f)
-    <div>Subject:
-      <ul class="errorlist"><li>This field is required.</li></ul>
-      <input type="text" name="subject" maxlength="100" required aria-invalid="true">
+    <div>
+      <label for="id_subject">Subject:</label>
+      <ul class="errorlist" id="id_subject_error"><li>This field is required.</li></ul>
+      <input type="text" name="subject" maxlength="100" required aria-invalid="true" aria-describedby="id_subject_error" id="id_subject">
     </div>
-    <div>Message:
+    <div>
+      <label for="id_message">Message:</label>
       <textarea name="message" cols="40" rows="10" required>Hi there</textarea>
     </div>
-    <div>Sender:
-      <ul class="errorlist"><li>Enter a valid email address.</li></ul>
-      <input type="email" name="sender" value="invalid email address" required aria-invalid="true">
+    <div>
+      <label for="id_sender">Sender:</label>
+      <ul class="errorlist" id="id_sender_error"><li>Enter a valid email address.</li></ul>
+      <input type="email" name="sender" value="invalid email address" maxlength="320" required aria-invalid="true" aria-describedby="id_sender_error" id="id_sender">
     </div>
-    <div>Cc myself:
-      <input type="checkbox" name="cc_myself" checked>
+    <div>
+        <label for="id_cc_myself">Cc myself:</label>
+        <input type="checkbox" name="cc_myself" id="id_cc_myself" checked>
     </div>
 
+If a custom ``aria-describedby`` is provided when defining the widget this will
+be used in place of the default value. Consideration will need to be given how
+to associate errors with their inputs to allow these descriptions to be read to
+screen reader users.
```

Comment:
This is rather horrid. Suggestions welcome. 

To help explain my concern -- if you added `aria-describedby` to your `widget.attrs` when defining the form that will be used in place of any thing that django's going to add. That is django won't add a reference to the errors. However, adding an additional `aria-describedby` only when the field has errors it a lot more logic for the user to write. I'm not sure I can explain all of that here, at least not succinctly.

Reply:
I'm not sure how we can explain that either, somehow it feels like we should solve this for them. At the very least we should provide a complete example about how to override `aria-describedby` in a correct manner? (I'm honestly not sure what would be the correct way to do this :-))

Reply:
As a ugly suggestion but to get us thinking, `Widget` can have something like
```python
    def get_aria_describedby(self, has_errors=False):
        return self.attrs.get("aria-describedby")
```
and we have in `build_widget_attrs`
```python
        widget_aria_describedby = self.widget.attrs.get_aria_describedby(
            has_errors=bool(self.errors)
        )
        if widget_aria_describedby:
            attrs["aria-describedby"] = widget_aria_describedby
```
Then `get_aria_describedby` can be documented and overwritten for custom widgets?

Reply:
With the new `aria_describedby` property and the upcoming easing of boundfield customisation I am much less concerned now. There will be a reasonable escape hatch if folk need it.

Reply:
You can put multiple, space-separated IDs in `aria-describedby`. We could append the error ID to the value already provided by the user.